### ISSUE TITLE
Render Helper w/out Sinatra

### ISSUE THREAD

**Issue by maxlinc** [padrino/padrino-framework#1872]
There are definitely parts of Padrino, particularly the helpers, that don't seem to be limited to Sinatra. Their usefulness certainly isn't limited to Sinatra and in many cases the implementation does not require Sinatra.

For example, Middleman, the popular static site generator uses [Padrino-based helpers](https://middlemanapp.com/basics/helpers/) so you can use those helpers for generating static sites. Grape has [already been mentioned](https://github.com/padrino/padrino-framework/issues/1854), and they could also be useful in Jekyll tag plugins.

The place I've been most interested in using them is in [Thor generators](https://github.com/erikhuda/thor/wiki/Generators). Those are the generators that power everything from "bundle gem" to "rails generate". I have generators that build documentation or HTML reports, so the Padrino helpers are useful.

I haven't had any trouble with most of the helpers - I've used the Output, Tag and Asset helpers without any problems, but I am not able to easily use the Render helper because I'll get a method missing error.

It seems that Middleman was able to support Padrino's render helper, but had to implement a [Sinatra/Padrino compatible render interface](https://github.com/middleman/middleman/blob/7c37d4ba5154dcd1bc93819752710b1c28b3f9b8/middleman-core/lib/middleman-core/core_extensions/rendering.rb#L195-L222).

I'm wondering if either:
- You could include assumptions in the helper documentation about the interface implemented by the class they are mixed into, like the required signature of `render` to use the RenderHelpers.

or

- Fallback to Tilt or some other approach if render is not implemented (i.e. not inside Sinatra or a Sinatra-compatible class).

**Comment by skade** [comment]
Another option would be to just implement the shim ourselves. I see the issue that Padrino basically provides the only proper view engine that is halfway independent of a framework in the Ruby space.

**Comment by nesquena** [comment]
Yeah this is an interesting point I've thought about before, there isn't much that actually ties the helpers module to Sinatra, they are just generally useful HTML generation methods for the most part (except for the rendering and partial logic). Thanks for bringing up this issue for us to think about.

**Comment by maxlinc** [comment]
I just switched some Thor code that looked like this:

```ruby
Crosstest.manifest.scenarios.values.each do |scenario|
  @scenario = scenario
   template 'templates/_test_report.html.tt', "details/#{scenario.slug}.html"
end
```

Over to this:
```ruby
template_file = find_in_source_paths("templates/_test_report.html.haml")
template = Tilt.new(template_file)
Crosstest.manifest.scenarios.values.each do |scenario|
  @scenario = scenario
  add_file "details/#{scenario.slug}.html" do
    template.render(self)
  end
end
```

It was interesting to see that since the Padrino helpers were loaded the template was an instance of Padrino::Rendering::HamlTemplate rather than Tilt::HamlTemplate. This change also means I'm not using Thor's template rendering at all, it's all Tilt/Padrino. The only thing I use from Thor is `find_in_source_paths`.

I can't use partials, though, because:
```
NoMethodError: super: no superclass method `render' for #<Crosstest::Command::Generate::Dashboard:0x007fc3b544ad38>
from /opt/boxen/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/padrino-helpers-0.12.4/lib/padrino-helpers/output_helpers.rb:33:in `render'
```

So I'm not sure about how the Sinatra stuff works, but in terms of pure-Tilt setup it seems like it would just be useful to customize the search paths for templates, but once Padrino finds the partial it seems like `Tilt.new(partial).render(locals)` would work rather than each Tilt-based framework having a render method.

Hope that makes sense... I wasn't exactly sure what sort of a shim you had in mind.
